verifiers/internal/gha/provenance_test.go

```diff
@@ -322,14 +326,37 @@ func Test_verifySourceURI(t *testing.T) {
 			expectedSourceURI: "https://github.com/some/repo",
 		},
 		{
-			name:               "match source no git no material ref",
-			provTriggerURI:     "git+https://github.com/some/repo@v1.2.3",
-			provMaterialsURI:   "git+https://github.com/some/repo",
-			allowNoMaterialRef: true,
-			expectedSourceURI:  "https://github.com/some/repo",
+			name:              "match source no git no material ref (npm)",
+			provBuildType:     common.NpmCLIBuildTypeV1,
+			provTriggerURI:    "git+https://github.com/some/repo@v1.2.3",
+			provMaterialsURI:  "git+https://github.com/some/repo",
+			expectedSourceURI: "https://github.com/some/repo",
+		},
+		{
+			name:              "mismatch source material ref (npm)",
+			provBuildType:     common.NpmCLIBuildTypeV1,
+			provTriggerURI:    "git+https://github.com/some/repo@v1.2.3",
+			provMaterialsURI:  "git+https://github.com/some/repo@v1.2.4",
+			expectedSourceURI: "https://github.com/some/repo",
+			err:               serrors.ErrorInvalidDssePayload,
+		},
+		{
+			name:              "match source no git no material ref (byob)",
+			provBuildType:     common.BYOBBuildTypeV0,
+			provTriggerURI:    "git+https://github.com/some/repo@v1.2.3",
+			provMaterialsURI:  "git+https://github.com/some/repo",
+			expectedSourceURI: "https://github.com/some/repo",
+		},
+		{
+			name:              "mismatch source material ref (byob)",
+			provBuildType:     common.BYOBBuildTypeV0,
+			provTriggerURI:    "git+https://github.com/some/repo@v1.2.3",
+			provMaterialsURI:  "git+https://github.com/some/repo@v1.2.4",
+			expectedSourceURI: "https://github.com/some/repo",
+			err:               serrors.ErrorInvalidDssePayload,
```

Comment:
We already have a test that verifies that `VerifyBranch` fails when the branch or ref is empty (i.e. when the provenance implementation returns an empty string).

I added a test to `BYOBProvenance` to test when the resolved dependency has no ref.

I guess the question is whether we want to allow something like `--source-branch ""` to pass in this case where `sha1` is different. Right now we don't allow it because the `Provenance` implementations (BYOB, etc.) return an error if the resolved dependencies don't have a ref, but we could just as easily allow it and return an empty string instead.

Reply:
I think I'll leave it as it is now since that's the closest to the current behavior. We can revisit it in the future if needed.